File: frontend/src/api/stock.ts
import { api } from "./client";

export interface KpisStock {
  nb_articles: number;
  disponible: number;
  rupture: number;
  alerte: number;
  valeur_stock: number;
}

export type TypeMouvement = "entree" | "sortie" | "ajustement" | "transfert";

export interface MouvementInput {
  article_id: number;
  type: TypeMouvement;
  quantite: number;
  cout_unitaire?: number | null;
  motif?: string | null;
  note?: string | null;
  magasin_destination_id?: number | null;
}

export interface Mouvement {
  id: number;
  article_id: number;
  article_designation?: string | null;
  magasin_id: number;
  magasin_destination_id?: number | null;
  type: TypeMouvement;
  quantite: number;
  cout_unitaire?: number | null;
  motif?: string | null;
  created_at: string;
}

export interface InventaireLigne {
  id: number;
  article_id: number;
  article_designation?: string | null;
  quantite_theorique: number;
  quantite_comptee: number;
  ecart: number;
}

export interface Inventaire {
  id: number;
  reference: string;
  magasin_id: number;
  statut: "brouillon" | "valide";
  note?: string | null;
  created_at: string;
  lignes: InventaireLigne[];
}

export async function kpisStock(): Promise<KpisStock> {
  return (await api.get<KpisStock>("/stock/kpis")).data;
}

export async function creerMouvement(input: MouvementInput): Promise<Mouvement> {
  return (await api.post<Mouvement>("/stock/mouvements", input)).data;
}

export async function listerMouvements(): Promise<Mouvement[]> {
  return (await api.get<Mouvement[]>("/stock/mouvements")).data;
}

export async function listerInventaires(): Promise<Inventaire[]> {
  return (await api.get<Inventaire[]>("/stock/inventaires")).data;
}

export async function creerInventaire(payload: {
  note?: string;
  lignes: { article_id: number; quantite_comptee: number }[];
}): Promise<Inventaire> {
  return (await api.post<Inventaire>("/stock/inventaires", payload)).data;
}

export async function validerInventaire(id: number): Promise<Inventaire> {
  return (await api.post<Inventaire>(`/stock/inventaires/${id}/valider`)).data;
}

File: backend/app/schemas/stock.py
"""Schémas Pydantic pour le stock, les mouvements et l'inventaire."""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class KpisStock(BaseModel):
    nb_articles: int
    disponible: int
    rupture: int
    alerte: int
    valeur_stock: Decimal


class MouvementCreate(BaseModel):
    article_id: int
    type: str = Field(description="entree | sortie | ajustement | transfert")
    quantite: Decimal
    cout_unitaire: Decimal | None = Field(default=None, ge=0)
    motif: str | None = Field(default=None, max_length=255)
    note: str | None = None
    magasin_destination_id: int | None = None


class MouvementOut(BaseModel):
    id: int
    article_id: int
    article_designation: str | None = None
    magasin_id: int
    magasin_destination_id: int | None = None
    type: str
    quantite: Decimal
    cout_unitaire: Decimal | None = None
    motif: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class StockLigneOut(BaseModel):
    article_id: int
    reference: str
    designation: str
    quantite: Decimal
    seuil_alerte: Decimal
    prix_achat_moyen: Decimal
    prix_vente: Decimal

    model_config = {"from_attributes": True}


# --- Inventaire --------------------------------------------------------------
class InventaireLigneIn(BaseModel):
    article_id: int
    quantite_comptee: Decimal = Field(ge=0)


class InventaireCreate(BaseModel):
    note: str | None = Field(default=None, max_length=255)
    lignes: list[InventaireLigneIn] = []


class InventaireLigneOut(BaseModel):
    id: int
    article_id: int
    article_designation: str | None = None
    quantite_theorique: Decimal
    quantite_comptee: Decimal
    ecart: Decimal

    model_config = {"from_attributes": True}


class InventaireOut(BaseModel):
    id: int
    reference: str
    magasin_id: int
    statut: str
    note: str | None = None
    created_at: datetime
    lignes: list[InventaireLigneOut] = []

    model_config = {"from_attributes": True}
